### DISCUSSION TITLE
Efficiency coloring scheme

### DISCUSSION THREAD

**Discussion by CapnBry** [stronnag/bbl2kml#19] [Q&A] [answered]
Hi strongnag! I've been using your bbl2kml converter for a bunch of RSSI testing and the values color code exactly as expected. However, I'm having a hard time understanding the coloring for efficiency. If I am looking at this section as an example (with `-gradient rdgn`)
![image](https://user-images.githubusercontent.com/677183/119370183-ff4def00-bc82-11eb-80fb-9d9eb1fecf43.png)

The first green point after the red is 55.7mah/km the red one just South of it is 54.3mah.km. Here are all the efficiencies of the green point to through all the red points and the first two semi-green ones and two greenyellow:
```
55.7 // green
54.3
52.6
51.4
50.1
52.2
52.6
54.5
54.5
55.6 // green
54.4
54.0
53.9
60.1 // semigreen
62.5 // semigreen
67.2 // greenyellow
68.8 // greenyellow
```

The average efficiency is 77.2mah/km so I don't think the range is being scaled down too much to make them just go full red or green. The efficiency numbers also don't seem to be wrong either. I've [checked the calculation](https://github.com/stronnag/bbl2kml/blob/751b123527930f106372f3ac47c455e1f684aebf/pkg/kmlgen/kmlbuilder.go#L108) and it seems right to me but maybe there's something I am missing about how the points are colored or something about how go works.

EDIT: Wait a second, are the Qval comparisons swapped? Like should the code be:
```
                        // Efficiency is mah/km so high numbers are bad (0)
			if r.Effic > qval1 {
				r.Qval = 0
			} else if r.Effic < qval0 {
				r.Qval = 100
			} else {
				r.Qval = 100*(1-(r.Effic-qval0)/(qval1-qval0))
			}
```

It would also be nice to know wh/km instead of mah/km considering this is a 2S1P 18560 pack so my mah/km are going to be like 30% higher at the end of the pack compared to the beginning due to the giant voltage difference.

I can provide the bbl file to you directly if needed.

**Comment by stronnag** [comment]
Thanks Capt. 

I'll have a look at this, the BBL would be appreciated as it'll save me scrapping through my contributed stash for good examples.

And yes, it does appear if the end points are swapped. RSSI was so much easier for my simple mind.

**Reply by CapnBry** [reply]
Here is the log file https://heatermeter.com/dl/tmp/LOG00031.zip

And looking at the code again I am certain the qval values are just swapped. Fully red is currently super good efficiency (low Effic) and fully green is super bad efficiency (high Effic). The points that are scaled are fine, just those 5% outliers have their colors swapped and that's what was throwing me off.

Swill would love it to be based on wh/km though if you have time :)

**Comment by stronnag** [comment]
Thanks for the log; indeed the endpoints were swapped. I'm looking at wh/km just now.

* You'd like the whole colour grading to be based on wh (via mah), rather than just a addtional value in the pop-up ?

**Reply by CapnBry** [reply]
Yeah the color grading would be ideal because I'd like to view the whole flight at once. Maybe swap them both (popup and coloring) like a osd_stats_energy_unit setting since I'm not sure folks would want to compare both wh/km and mah/km at once.

I'm sorry if I posted this to the wrong place initially. I actually thought I just wasn't understanding the colors after randomly clicking a few dozen samples and not seeing the pattern, but once I typed all that out it started making sense that something was wrong.

**Comment by stronnag** [comment, marked as answer]
OK, so there is now a new option:
```
 -energy-unit string
    	Energy unit [mah, wh] (default "mah")
```
and if one always wants wh, set it in the config file, the command line will override for exceptional use. It would be possible to have both in the KMZ,though it appears big enough already.

The current code in the `development` branch generates a log to STDERR if the environment variable `DEBUG_EFFIC` is defined (to anything), thusly:
```
DEBUG_EFFIC=1 ./flightlog2kml -gradient rdgn -energy-unit=wh LOG00031.TXT  2>/tmp/e-wh.txt
```
I'll remove the debug log prior to release, but if want to look at it, either pull and `go build` development (or tell me for which OS you'd like a binary). 
Examples attached. Not sure there's that much difference in colouration.
[log31-out.zip](https://github.com/stronnag/bbl2kml/files/6541465/log31-out.zip)

**Reply by CapnBry** [reply]
I'll have to admit I have been dragging my feet about installing go because I thought I didn't need another 10GB environment cluttering up my poor SSD but golly if it isn't small and one click to install.

Looks perfect though. I think what was throwing me off was that there were green sections before with terrible efficiency and red sections with great efficiency and it made me question the whole system. Now it looks like I'd expect, and I can clearly see where the extremes were just flipped. I've spot checked many of the outliers and mid-range points and the coloring is 100% accurate now. Good choice on making the energy and efficiency numbers just report both-- saving a lot of KML overhead vs separate streams. It all looks good to me and I've tested it with several different logs from different locations. They all look perfect.

Thanks so much, you're my favorite person as always! 🤩 Also, sorry about the message spam this morning. github kept telling me the comment could not be saved so I just kept clicking REPLY.

**Comment by stronnag** [comment]
Bryan,

Thanks for testing.

Aside: I first tried Go when it came out in 2009; even submitted some patches to the then broken stdlib; got fed up with the ever changing beta APIs. Moved on. Came back to it recently. Cross-compile for a zillion platforms and it'll just work. Static binaries so no external dependencies. Love it.

-jonathan
